File: packages/flutter-im-ui/test/theme_mode_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

/// The shared theme table (`spec/theme-mode-vectors.json`): the host holds the person's choice and the kit
/// resolves it. The same file is read by the Vue, SwiftUI and Compose tests.
void main() {
  final table = jsonDecode(File('../../spec/theme-mode-vectors.json').readAsStringSync()) as Map<String, dynamic>;
  final cases = (table['cases'] as List).cast<Map<String, dynamic>>();

  FlareThemeMode modeOf(String name) => FlareThemeMode.values.firstWhere((value) => value.name == name);

  test('the table covers every mode', () {
    expect(cases.length, FlareThemeMode.values.length * 2);
  });

  for (final vector in cases) {
    test('theme vector: ${vector['id']}', () {
      expect(
        flareThemeIsDark(modeOf(vector['mode'] as String), systemDark: vector['systemDark'] as bool),
        vector['dark'],
      );
    });
  }

  testWidgets('the colours a body draws follow the mode the host chose, not the platform', (tester) async {
    tester.platformDispatcher.platformBrightnessTestValue = Brightness.light;
    addTearDown(tester.platformDispatcher.clearPlatformBrightnessTestValue);
    late FlareColors underSystem;
    late FlareColors underDark;
    await tester.pumpWidget(MaterialApp(
      home: Column(children: [
        FlareTheme(child: Builder(builder: (context) {
          underSystem = FlareColors.of(context);
          return const SizedBox.shrink();
        })),
        FlareTheme(mode: FlareThemeMode.dark, child: Builder(builder: (context) {
          underDark = FlareColors.of(context);
          return const SizedBox.shrink();
        })),
      ]),
    ));
    expect(underSystem.bgPrimary, FlareColors.violetLight.bgPrimary);
    expect(underDark.bgPrimary, FlareColors.violetDark.bgPrimary);
  });

  testWidgets('system follows the app, which follows the platform, and the strings carry the three labels', (tester) async {
    tester.platformDispatcher.platformBrightnessTestValue = Brightness.dark;
    addTearDown(tester.platformDispatcher.clearPlatformBrightnessTestValue);
    late FlareColors drawn;
    // The app resolves `system` the way every Flutter app does — light and dark themes with
    // `ThemeMode.system` — and the kit follows the theme it lands on.
    await tester.pumpWidget(MaterialApp(
      theme: ThemeData(brightness: Brightness.light),
      darkTheme: ThemeData(brightness: Brightness.dark),
      themeMode: ThemeMode.system,
      home: FlareTheme(child: Builder(builder: (context) {
        drawn = FlareColors.of(context);
        return const SizedBox.shrink();
      })),
    ));
    expect(drawn.bgPrimary, FlareColors.violetDark.bgPrimary);
    const strings = FlareStrings();
    expect([strings.themeSystem, strings.themeLight, strings.themeDark], ['跟随系统', '浅色', '深色']);
  });
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/ThemeModeVectorsTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The shared theme table (`spec/theme-mode-vectors.json`): the host holds the person's choice and the kit
 * resolves it. The same file is read by the Vue, Flutter and SwiftUI tests.
 */
class ThemeModeVectorsTest {
    private fun vectorsFile(): File {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "spec/theme-mode-vectors.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile
        }
        error("spec/theme-mode-vectors.json not found")
    }

    @Test fun everyCaseResolvesTheSameWayHere() {
        val cases = (FlareJson.parse(vectorsFile().readText()) as Map<*, *>)["cases"] as List<*>
        assertEquals(FlareThemeMode.entries.size * 2, cases.size, "the shared table lost cases")
        for (raw in cases) {
            val case = raw as Map<*, *>
            val id = case["id"] as String
            val mode = FlareThemeMode.entries.first { it.name.equals(case["mode"] as String, ignoreCase = true) }
            assertEquals(case["dark"], flareThemeIsDark(mode, case["systemDark"] as Boolean), id)
        }
    }

    @Test fun theStringsCarryTheThreeLabels() {
        val strings = FlareStrings()
        assertTrue(listOf(strings.themeSystem, strings.themeLight, strings.themeDark).none { it.isBlank() })
        assertEquals(listOf("跟随系统", "浅色", "深色"), listOf(strings.themeSystem, strings.themeLight, strings.themeDark))
    }
}
